### PR TITLE
[PyTorch][Training][EC2] PT 2.1 Currency

### PR DESCRIPTION
**Pull request by sirutBuasai**
*GitHub Issue #, if available:*

**Note**: 
- If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

- All PR's are checked weekly for staleness. This PR will be closed if not updated in 30 days.

### Description
Added folder structure for PT 2.1 Inference (buildspecs, Dockerfiles).
In addition:

- Disabled torchdata tests because torchdata has stopped active development (https://github.com/pytorch/data/#torchdata-see-note-below-on-current-status). The corresponding torchdata version for PT 2.1.0 is 0.7.0, which has been published on pip and conda, but the release tag on github is missing (only RC tag), which is causing test failures. If torchdata never creates the release tag, we may have to update our torchdata test logic to accommodate.
- Removed pillow version pin due to resolution of https://github.com/pytorch/vision/issues/7660
- Removed commented Horovod sections
- Added ffmpeg from our conda channel due to https://github.com/aws/deep-learning-containers/issues/3069
- Updated [transformer engine](https://github.com/NVIDIA/TransformerEngine/tree/v0.13) to latest v0.13
- Updated [flash attention](https://github.com/Dao-AILab/flash-attention/tree/v2.0.4) to v2.0.4 (latest compatible version with transformer engine)
- Updated buildspec.yml to PT2.1 and build cpu, cu118, cu121
- Updated Python 3.10 patch version to latest 3.10.12
- Updated Mamba version to latest 23.3.1-1
- Updated EFA version to latest 1.27.0
- Updated NCCL version to 2.18.5
- Updated Triton version to 2.1.0
- Updated CuDNN 8.7.0.84 for cu118 and 8.9.2.26 for cu121
- Updated Conda channel for cpu container
- Pin `cryptography>=41.0.4` due to CVE patching
- Pin `accelerate==0.22.0` due to `huggingface_hub` dependency

### Tests run

Commit [6250b89](https://github.com/aws/deep-learning-containers/pull/3389/commits/6250b89c73adad5fabaf350f270f667b04d0c8f8)
- cu118, cu121, cpu Build
- cu118, cu121, cpu EC2 Test
- cu118, cu121, cpu EKS Test
- cu118, cu121, cpu Sanity Test

Commit [e7761b9](https://github.com/aws/deep-learning-containers/pull/3389/commits/e7761b94f08d7fd7b51274c5a800ee6840d77f79)
- cu118, cu121, cpu ECS Test

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+# install PyTorch related python packages (depends on PyTorch)
+RUN /opt/conda/bin/mamba install -y -c conda-forge \
+  # needed by fastai.distributed
+  # pin accelerate to 0.22.0 as newer version gives error "0.23.0 requires huggingface-hub, which is not installed."
+  accelerate=0.22.0 \
+  && /opt/conda/bin/mamba install -y -c fastai fastai \
+  # fastai conda strictly depends on dataclasses
+  && pip uninstall -y dataclasses \
+  && pip install torchtnt \
+# explicitly install triton build-time dependencies, which won't get installed by conda for some reason
```
Comment (junpuf) [inline pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu:308]: remove this part about "explicitly install triton build-time dependencies", as have our torchtriton already.

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu [hunk 2]
```diff
@@ -0,0 +1,321 @@
+ARG PYTHON=python3
```
Comment (junpuf) [inline pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:1]: some comments from cu118 dockerfile also apply to this one

File: test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py [hunk 3]
```diff
@@ -583,6 +583,7 @@ def test_pytorch_training_torchaudio_cpu(
     execute_ec2_training_test(ec2_connection, pytorch_training, PT_TORCHAUDIO_CMD)
```
Comment (junpuf) [inline test/dlc_tests/ec2/pytorch/training/test_pytorch_training.py:583]: we might want to add a new test to ensure our packages like pytorch, torchvision, torchaudio, torchtriton, etc are not being replaced

**Review by junpuf — COMMENTED** [review]
Added some comments

File: pytorch/training/docker/2.1/py3/Dockerfile.cpu [hunk 4]
```diff
@@ -0,0 +1,327 @@
+ARG PYTHON=python3
+ARG PYTHON_VERSION=3.10.12
+ARG MAMBA_VERSION=23.3.1-1
+ARG PYTORCH_VERSION_EC2=2.1.0
+ARG PYTORCH_VERSION_SM=2.1.0
+ARG EC2_CONDA_CHANNEL=https://aws-ml-conda-ec2.s3.us-west-2.amazonaws.com
+ARG SM_CONDA_CHANNEL=https://aws-ml-conda-sm.s3.us-west-2.amazonaws.com
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/Dockerfile.cpu]: consolidated these into `CONDA_CHANNEL` in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 5]
```diff
@@ -0,0 +1,658 @@
...
+# PyTorch Binaries
+ARG PT_EC2_TRAINING_URL
+ARG CONDA_CHANNEL
+ARG NCCL_ASYNC_ERROR_HANDLING
+
+WORKDIR /
+
+# Install ec2 PyTorch
+RUN /opt/conda/bin/mamba install -y python=${PYTHON_VERSION} pytorch=${PYTORCH_VERSION_EC2} \
+  pytorch-cuda=11.8 cuda-nvcc=11.8.89 torchvision torchaudio torchtext aws-ofi-nccl-dlc ffmpeg \
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu]: removed explicit mamba install ffmpeg in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+
+# Copy workaround script for incorrect hostname
+COPY changehostname.c /
+COPY start_with_right_hostname.sh /usr/local/bin/start_with_right_hostname.sh
+
+RUN chmod +x /usr/local/bin/start_with_right_hostname.sh
+
+WORKDIR /root
+
+RUN git clone https://github.com/Dao-AILab/flash-attention.git \
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu]: fix flash-attn installation and added transformer engine installation in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+    && cd csrc/fused_dense_lib && pip install . && cd ../../ \
+    && cd csrc/ft_attention && pip install . && cd ../../ \
+    && cd .. && rm -rf flash-attention
+
+RUN pip install --no-cache-dir -U \
+    smclarify \
+    "sagemaker>=2,<3" \
+    "sagemaker-experiments<1" \
+    sagemaker-pytorch-training \
+    flash-attn==0.2.8 \
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu]: removed in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+    triton==2.1.0
```
[the remark this answers is not in the document]
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu]: removed in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu (hunk 1 above)
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu]: removed in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu (hunk 2 above)
Comment (sirutBuasai) [reply pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:1]: fixed alongside cu118 file in [af40080](https://github.com/aws/deep-learning-containers/pull/3389/commits/af4008083c72013e091f6da1ee0aa62117361752)

File: pytorch/training/docker/2.1/py3/Dockerfile.cpu [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    ipython \
+    # Adding package for studio kernels
+    ipykernel \
+  && /opt/conda/bin/mamba install -y -c dglteam dgl \
+  && /opt/conda/bin/mamba clean -afy
+
+RUN pip install --no-cache-dir --upgrade pip --no-cache-dir --trusted-host pypi.org --trusted-host files.pythonhosted.org \
+ && ln -s /opt/conda/bin/pip /usr/local/bin/pip3
+
+RUN curl -o /license.txt https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.0/license.txt
```
Comment (sirutBuasai) [inline pytorch/training/docker/2.1/py3/Dockerfile.cpu:150]: should we create a new dir for `pytorch-2.1`?

File: pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+
+RUN rm -rf /root/.ssh/ && \
+ mkdir -p /root/.ssh/ && \
+ ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
+ cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+ && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
+
+RUN curl -o /license.txt https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.0/license.txt
```
Comment (sirutBuasai) [inline pytorch/training/docker/2.1/py3/cu121/Dockerfile.gpu:199]: should we create a new dir for `pytorch-2.1`?

File: pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+
+RUN rm -rf /root/.ssh/ && \
+ mkdir -p /root/.ssh/ && \
+ ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
+ cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys \
+ && printf "Host *\n StrictHostKeyChecking no\n" >> /root/.ssh/config
+
+RUN mkdir -p /etc/pki/tls/certs && cp /etc/ssl/certs/ca-certificates.crt /etc/pki/tls/certs/ca-bundle.crt
+
+RUN curl -o /license.txt https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.0/license.txt
```
Comment (sirutBuasai) [inline pytorch/training/docker/2.1/py3/cu118/Dockerfile.gpu:192]: should we create a new dir for `pytorch-2.1`?

File: pytorch/training/docker/2.1/py3/Dockerfile.cpu (hunk 9 above)
Comment (junpuf) [reply pytorch/training/docker/2.1/py3/Dockerfile.cpu:151]: use `https://aws-dlc-licenses.s3.amazonaws.com/pytorch-2.1/license.txt`

**Review by junpuf — APPROVED** [review]
LGTM

**Pull request merged**